File: dotnet/Zirve.Sdk/Db/DbManager.cs
using System;
using System.Data;
using System.Threading.Tasks;
using Npgsql;
using Zirve.Sdk.Config;

namespace Zirve.Sdk.Db;

/// <summary>
/// Zirve Db Manager — PostgreSQL Integration.
/// Utilizes Npgsql for pooling, tenant isolation, and connection management.
/// </summary>
public class DbManager : IAsyncDisposable
{
    private readonly NpgsqlDataSource _dataSource;
    private readonly string _connectionString;

    public DbManager(ConfigManager configManager)
    {
        var cfg = configManager.Module("db");
        var host = cfg.GetValueOrDefault("host", "localhost");
        var port = cfg.GetValueOrDefault("port", "5432");
        var dbname = cfg.GetValueOrDefault("dbname", "zirve");
        var user = cfg.GetValueOrDefault("user", "postgres");
        var password = cfg.GetValueOrDefault("password", "");

        // Build ADO.NET Connection String with pooling logic suitable for high concurrency
        var builder = new NpgsqlConnectionStringBuilder
        {
            Host = host,
            Port = int.TryParse(port, out var p) ? p : 5432,
            Database = dbname,
            Username = user,
            Password = password,
            Pooling = true,
            MinPoolSize = 1,
            MaxPoolSize = 100,
            Timeout = 3
        };

        _connectionString = builder.ToString();
        _dataSource = NpgsqlDataSource.Create(_connectionString);
    }

    /// <summary>
    /// Executes a single SQL statement that returns no data (e.g. INSERT, UPDATE).
    /// </summary>
    public async Task<int> ExecuteAsync(string sql, params NpgsqlParameter[] parameters)
    {
        await using var cmd = _dataSource.CreateCommand(sql);
        cmd.Parameters.AddRange(parameters);
        return await cmd.ExecuteNonQueryAsync();
    }

    /// <summary>
    /// Executes a SQL query and returns the first column of the first row.
    /// </summary>
    public async Task<T?> ScalarAsync<T>(string sql, params NpgsqlParameter[] parameters)
    {
        await using var cmd = _dataSource.CreateCommand(sql);
        cmd.Parameters.AddRange(parameters);
        var res = await cmd.ExecuteScalarAsync();
        if (res == null || res == DBNull.Value) return default;
        return (T)res;
    }

    /// <summary>
    /// Gets a fully managed DbConnection intended for passing to an ORM like Dapper.
    /// Make sure to configure the schema appropriately when multi-tenant.
    /// </summary>
    public NpgsqlConnection GetConnection()
    {
        return _dataSource.CreateConnection();
    }

    /// <summary>
    /// Runs operations inside a database transaction safely.
    /// Auto commits or rollbacks on Exception.
    /// </summary>
    public async Task<T> TransactionAsync<T>(Func<NpgsqlTransaction, NpgsqlConnection, Task<T>> callback)
    {
        await using var connection = await _dataSource.OpenConnectionAsync();
        await using var transaction = await connection.BeginTransactionAsync(IsolationLevel.ReadCommitted);
        
        try
        {
            var result = await callback(transaction, connection);
            await transaction.CommitAsync();
            return result;
        }
        catch
        {
            await transaction.RollbackAsync();
            throw;
        }
    }

    /// <summary>
    /// Changes the PostgreSQL schema context dynamically. Useful for strictly isolated tenants.
    /// </summary>
    public async Task SetTenantAsync(NpgsqlConnection connection, string tenantId)
    {
        var safeId = string.Concat(tenantId.Split(System.IO.Path.GetInvalidFileNameChars()));
        await using var cmd = connection.CreateCommand();
        cmd.CommandText = $"SET search_path TO tenant_{safeId.Replace("'", "")}, public";
        await cmd.ExecuteNonQueryAsync();
    }

    /// <summary>
    /// Basic health check simply executing a SELECT 1 query to database.
    /// </summary>
    public async Task<bool> HealthAsync()
    {
        try
        {
            await using var cmd = _dataSource.CreateCommand("SELECT 1");
            await cmd.ExecuteScalarAsync();
            return true;
        }
        catch
        {
            return false;
        }
    }

    public async ValueTask DisposeAsync()
    {
        if (_dataSource != null)
        {
            await _dataSource.DisposeAsync();
        }
    }
}

File: typescript/src/Db/DbManager.ts
import { Pool, PoolClient, QueryResult } from 'pg';

/**
 * Zirve DB Manager — PostgreSQL & MariaDB.
 * 
 * Provides prepared queries, multi-tenant schema switching, and transactions.
 * Utilizes generic `pg` Pool for connection management.
 */
export class DbManager {
  private pool: Pool;
  private config: Record<string, any>;

  constructor(config: Record<string, any>) {
    this.config = config;
    this.pool = new Pool({
      host: config.host || 'localhost',
      port: Number(config.port) || 5432,
      database: config.dbname || 'zirve',
      user: config.user || 'postgres',
      password: config.password || '',
      max: 20, // max connection pool size
      idleTimeoutMillis: 30000,
    });
  }

  /**
   * Execute a parameterized SQL query.
   */
  public async query<T = any>(sql: string, params: any[] = []): Promise<T[]> {
    const result = await this.pool.query(sql, params);
    return result.rows;
  }

  /**
   * Execute a single scalar query, returning the first column of the first row.
   */
  public async scalar<T = any>(sql: string, params: any[] = []): Promise<T | null> {
    const rows = await this.query(sql, params);
    if (!rows || rows.length === 0) return null;
    return Object.values(rows[0])[0] as T;
  }

  /**
   * Safely execute operations within a database transaction.
   * Auto-commits on success, rolls back on exception.
   */
  public async transaction<T>(callback: (client: PoolClient) => Promise<T>): Promise<T> {
    const client = await this.pool.connect();
    try {
      await client.query('BEGIN');
      const result = await callback(client);
      await client.query('COMMIT');
      return result;
    } catch (err) {
      await client.query('ROLLBACK');
      throw err;
    } finally {
      client.release();
    }
  }

  /**
   * Set PostgreSQL `search_path` to isolated tenant schema for multi-tenant data.
   */
  public async tenant(tenantId: string): Promise<void> {
    // Escaping identifier to prevent SQL injection in schema name
    const cleanId = tenantId.replace(/[^a-zA-Z0-9_]/g, '');
    await this.query(`SET search_path TO tenant_${cleanId}, public`);
  }

  /**
   * Health check to ping the database.
   */
  public async health(): Promise<boolean> {
    try {
      await this.query('SELECT 1');
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Close the connection pool gracefully.
   */
  public async close(): Promise<void> {
    await this.pool.end();
  }
}
